### ISSUE TITLE
Deleting original js/css files from web accessible folder after compression

### ISSUE THREAD

**Issue by nirvana-msu** [yiisoft/yii2#9708]
Would be nice if `asset/compress` command deleted original js/css files that are published to web accessible folder, once compression is performed. They may contain some sensitive information (e.g. extensive comments), and not exposing them is one of the reasons for compression in the first place.

Even though these files are not served to the client browser directly, they are accessible if URL is known. And all you need to know is an 8-symbol hash, which is pretty easy to brute-force, and a relative path/filename, which quite often is easy to guess (e.g. I bet most projects have `/css/main.css`, and sometimes JS file names are similar to function names that you can find in a compressed script).

Are there any reasons not to delete original js/css files that have been compressed? Should be pretty easy to do.

**Comment by pana1990** [comment]
@nirvana-msu you may put js/css files in another directory(e.g. assets/js, assets/css)

**Comment by nirvana-msu** [comment]
Not sure what you mean by that. I don't have any problem with compressing or serving these files. What bothers me is that original uncompressed files are left in the web accessible directory after publishing (which is the first step during compression).

**Comment by pana1990** [comment]
@nirvana-msu I'm sorry, I had not understood.

**Comment by pana1990** [comment]
@nirvana-msu  can you check my pull request? 

https://github.com/yiisoft/yii2/pull/9826

**Comment by nirvana-msu** [comment]
@pana1990 Thanks for the pull request, but it's a little bit more complex than that. You cannot simply delete the whole bundle directory that was published, as apart from `js` / `css` files that were compressed, there may also be e.g. images that are still required for web pages. The only way is to iterate over each `js` / `css` file that got compressed and delete them (and only them) one by one.

**Comment by pana1990** [comment]
@nirvana-msu ok, i will review my code, thanks for your help :)

**Comment by pana1990** [comment]
@nirvana-msu check my pull request again, please

**Comment by nirvana-msu** [comment]
@pana1990 thanks. I checked pull request. It's better but it still has issues. What you do now is simply deleting all `js`/`css` files you find within the published directory. This is wrong, you need to delete *only* those files that were explicitly listed as bundle dependencies (and were added to compressed script). Otherwise you end up removing files which are still required for the page to function.

Here's a particular example where it fails. Below is a piece of code from my compressor template. Note that `yii\bootstrap\BootstrapAsset` is one of the bundles being compressed.
```php
    'jsCompressor' => 'java -jar compiler.jar --js {from} --js_output_file {to}',
    'cssCompressor' => 'java -jar yuicompressor-2.4.7.jar --type css {from} -o {to}',
    'bundles' => [
        'yii\web\JqueryAsset',
        'yii\web\YiiAsset',
        'yii\jui\JuiAsset',
        'yii\bootstrap\BootstrapAsset',
        ...
    ],
```

When the page is rendered, I see the following script included. Loading obviously fails (resulting in page not functioning correctly) because `bootstrap.js` is removed.
```php
<script src="/assets/28e9efd1/js/bootstrap.js"></script>
```

Debugging shows me that `yii\bootstrap\Dropdown` (which is called from a NavBar) registers `yii\bootstrap\BootstrapPluginAsset` bundle, which has the same `sourcePath` (and `basePath`) as `yii\bootstrap\BootstrapAsset`, and publishes `bootstrap.js` to the same directory.
```php
// inside yii\bootstrap\Dropdown->run()
BootstrapPluginAsset::register($this->getView());
```

Note that `yii\bootstrap\BootstrapAsset` is listed in my compression script because I explicitly specify it as a dependency in one of my other asset bundles, but `yii\bootstrap\BootstrapPluginAsset` is not listed. I did not even know it existed before today. It is clearly unreasonable to expect me to list all the bundles which share the same path - god knows how many of those are and where to find them.

Another example is `yii\widgets\ActiveFormAsset` which has the same `sourcePath` and `basePath` as `yii\web\YiiAsset`. And while the latter is among the bundles to compress, the former is not as I never deal with it directly in my code. This results in the following script not being able to load (and forms not functioning correctly)

```php
<script src="/assets/ecca149b/yii.activeForm.js"></script>
```

**Comment by pana1990** [comment]
@nirvana-msu thanks for your help, i think it works now, can you try it again?

**Comment by nirvana-msu** [comment]
@pana1990 yes, seems to work exactly as it should now. Hopefully that can be merged into one of the next releases.

Thanks!

**Comment by pana1990** [comment]
@nirvana-msu thank you for your help ;)

**Comment by klimov-paul** [comment]
The whole feature does not sounds good for me.
Deleting js/css files after compression eliminates the ability to use uncopressed asset bundles.
Switching back to uncompressed assets can be useful even at production server - mainly for debug purposes.
While those files remains such switching is just a matter of single row inside the config file. If original files are removed disabling the compressed AssetBundle will cause a broken 'front end'. This happens because uncompressed CSS and JS files are no longer exist, while the asset source folder is published (it contains images, fonts and so on). Thus original AssetBundles will not be republished and source CSS and JS files will be missing.

**Comment by nirvana-msu** [comment]
@klimov-paul obviously the assumption is that you will not be switching to debug version in production. It is just common sense to not even have such an ability.

You can switch to debug in your dev environment, or even on a staging server - all you need is to delete published directory and reload a page, or alternatively add `forceCopy` option which, btw, is also just a one line change in config file. That's not really hard to do, and you don't need that often.

Primary concern is to not expose any sensitive information, which is not the case as long as original files are accessible.

If, for some reason, you still insist on preserving the current behaviour, we can simply make this feature optional.

**Comment by klimov-paul** [comment]
> Primary concern is to not expose any sensitive information, which is not the case as long as original files are accessible.

I don't get it. What harm can cause ability to access to uncompresed JS or CSS file? Thier content already available inside compressed files. Besides guessing of the excact root for the published asset directory is a bit of challange. Also remember if you are compressing assets, which are not published, but accessed via `baseUrl` directly - they will remain at place one way or another.

**Comment by nirvana-msu** [comment]
I described the reasons in my first post, and also mentioned an easy way one could find some of the compressed folders. To re-state it again, main reason is stripping out comments. It is convenient to extensively document the code, and that may easily contain sensitive information. Besides, by minifying source only you can at least get the kids off messing around with it. Obviously it won't stop anyone who really wants to read it, but *every little helps* - you want to make the attacker's life as hard as possible using whatever you have at your disposal. There are plenty discussions on the web whether or not it makes sense, probably not worth having another one here.

Regarding `baseUrl` - all the sources I care about (specifically, those having potentially sensitive comments) are intentionally placed in a non web-accessible folder.

**Comment by klimov-paul** [comment]
Resolved by commit 56cc97ae420073ff4086db4119faacfbb610263d
